README.md
# Software Defined Radio - DAB+ (libsdrdab), version 3.0 beta

It is implementation of DAB/DAB+ software decoder. Input samples could be retrived from file or tuner (librtlsdr). Output sound could be save to file or played via GStreamer.

Developed by students of [AGH-UST](http://www.agh.edu.pl/en) in [Department of Telecommunications](http://kt.agh.edu.pl/en). 

Project webpage [sdr.kt.agh.edu.pl](https://sdr.kt.agh.edu.pl)


## Build and Run

Tested on Ubuntu 14.x, 16.x and other. Need: gcc and a few standard libraries. libsdrdab has it's own implementation of librtlsdr. Do not need librtlsdr, do not conflict with installed librtlsdr.

* download code 
* unpack
* install dependencies: ./sdrtool install-deps
* build: ./sdrtool build Release
* data: ./sdrtool download-data-light					# optional for IQ raw data
* run: 
*  ./build/bin/sdrdab-cli 						# from SDR USB tuner
*  ./build/bin/sdrdab-cli --open-file=./data/*.raw			# default is --file-type=raw
*  ./build/bin/sdrdab-cli --open-file=./data/*.raw --file-type=uint8	# or --file-type=raw
*  ./build/bin/sdrdab-cli --open-file=./data/*.dat --file-type=float


library is located in ./build/src/sdrdab/libsdrdab.a
cli-reference-implemenation is located in ./build/bin/sdrdab-cli

Installation is not supported so far. Experimental package building is supportet via ./sdrtool

## UnitTests

UnitTests may be useful for developing purpose. Steps necessary to perform it:

* ./sdrtool download-ut     # download raw data
* ./sdrtool matlab          # generate input and output for unittests (require Matlab(C) version 2006+)
* ./sdrtool build Unittest  # build library with unittests
* ./sdrtool run-unittests   # perform all unittests

### Matlab

Unittests includes DAB/DAB+ decoder (in Matlab) written by Michael Hoin [ZHAW Zürcher Hochschule für Angewandte Wissenschaften](https://www.zhaw.ch/) and extended by Tomasz Zielinski [AGH-UST Telecommunications Department](http://kt.agh.edu.pl/en). This code is used to generate input and output (truth) data for unittests. You need [Matlab](https://in.mathworks.com/products/matlab.html) to perform unittests.

### Data

It is possible to download some data (./sdrtool download-*) necessary for unittest and helpful for testing library. It consists of recordings come from publicly available and uncoded signal. Copyrights to pieces registered at that time belong to their authors. Recordings were not modified (apart from format conversion). Recordings can only be used in order to test the library; in particular further publishing or public reproducing is not allowed. 

## ToDo

### DAB(+) decoder - feature/algorithms (unsorted):

* debug AGC
* TEQ - time equalizer
* resampling (fs) made by means of remodulate+FEQ+interpolation after FFT
* custom resampling (fs), only necessary parts (like FFT in MSC), remove RingBuffer
* resampling == own thread? CPU vector (SSE/NEON) optimization?
* implementation of convolutional code simpler than Viterbi decoder (hard-decision, BCJR)
* smart algorithm for convolutional decoder choise (based on SNR?), independent FIC and MSC?
* USRP supports (input signal), other hardware?
* extend input file support: automatic (smart) detection
* decompress to raw samples (get rid of GStreamer and other framework), make resampling in time domain
* decoding extra data (images, etc…)
* RS smart on/off (depends on previous errors, CRC and firecode)
* support Rayleigh fading channels
* memory optimization (smaller buffer, adaptive buffers?)
* improve system delay (scheduler optimization), system should maintain 200-500 ms delay
* FIC SNR improvement: CIF’s repeats (energy could be accumulated before Viterbi and thus improve SNR)
* decode many (all) audio streams simultaneously
* FIle2FIle should not skip frames from beginning of stream
* test and tune to mobile case (Rayleigh channel)
* rewrite GUI (with vizualization: spectrum, SNR, errors, extra data (images), etc...)
* merge arm-development into development (extra CMake, stats generator, performance improvements, Neon instructions support, psd and snr estimation)
* performance (time) test integrated with build (per method on release build)
* performance (erorr) test integrated with build - parameters: SNR, fc, fs
* ARM (RaspberyPI v2)
* ARM Android NDK/SDK
* x86 Windows
* x86 MacOS: more test, build is already done
* refactoring DataFeeder (too complex)
* refactoring AudioDecoder (too complex?)
* all code (including library) on LGPL license
* integration with vlc/mplayer/kino/kodi plugin/other media player
* recover broken UnitTest
* improve Synchronizer (chose algorithm, remove or repair dependencies for remodulation)
* test synchronizer for weak signal (+-10 levels for 8-bit input sample)
* show information from NULL decoding (transmitter ID)
* software MIMO (check if it's possible to use two or more RTL dongle, synchronize and use MIMO algorithms)
* calculate SNR from PR (3rd algorithm for SNR calculation)
* analyze algorithm for SNR calculation from Demodulator
* improve fs and fc drifts
* implement time signatures/counter for signal in all classes (allows for precise synchronization)
* auto-scanning, search for all allowed frequency for DAB/DAB+ signal
* add CRC verification for XPAD (text)
* improve XPAD decoding (images)


### build (unsorted)

* performance (time) test integrated with build (per method on release build)
* performance (error) test integrated with build - parameters: SNR, fc, fs
* DataFeeder refractoring (too complex)
* AudioDecoder refractoring (too complex)
* package for Linux (Ubuntu and other distros)
* all code (including library) on LGPL license
* ARM (RaspberyPI v2)
* ARM Android NDK/SDK
* x86 Windows
* x86 MacOS: more test, build is already done
* integration with vlc/mplayer/kino/kodi plugin/other media players
* merge arm-development (2015) into development (extra CMake, stats generator, performance improvements, Neon instructions support, psd and snr estimation)


### DAB(+) encoder

* like [eti-tools](https://github.com/piratfm/eti-tools)
* supports USRP (real-time)
* supports simulation of various disturbances: AWGN, NBI, fading channels, rayleigh fading channels, etc...

## copyright

2017:
* Tomasz P. Zieliński (tzielin@agh.edu.pl)
* Jarosław Bułat (kwant@agh.edu.pl)          <---- contact person
* Konrad Bańka
* Szymon Bar
* Marcin Chroń
* Szymon Flakus
* Dominik Koza
* Piotr Kmiecik
* Krystian Majer
* Bartłomiej Ogorzałek
* Michał Piech
* Grzegorz Skołyszewski
* Mirosław Szewczyk
* Miłosz Śliwiński
* Marek Winiarski
* Mateusz Ziarko

2016:
* Tomasz P. Zieliński (tzielin@agh.edu.pl)
* Jarosław Bułat (kwant@agh.edu.pl)          <---- contact person
* Jakub Bernady
* Szymon Czubak
* Szymod Dawidow
* Alicja Gęgotek
* Joanna Gniadek
* Michał Jurczak
* Paweł Kazimierowicz
* Przemysław Piątek
* Anna Radoszek
* Marek Rzepecki
* Mikołaj Suliga
* Bartosz Stopa
* Marcin Trebunia
* Piotr Wierzbicki
* Adrian Włosiak
* Łukasz Włosowicz
* Mateusz Wnętrzak
* Kacper Żuk

2015:
* Tomasz P. Zieliński (tzielin@agh.edu.pl)
* Jarosław Bułat (kwant@agh.edu.pl)          <---- contact person
* Michał Babiuch (babiuch.michal@gmail.com)
* Ernest Biela
* Szymon Dąbrowski
* Piotr Jaglarz
* Adrian Karbowiak
* Sebastian Leśniak
* Rafał Palej
* Kacper Patro
* Michał Rybczyński
* Dawid Rymarczyk
* Michał Rzepka (mrzepka@student.agh.edu.pl)
* Krzysztof Szczęsny (kaszczesny@gmail.com)
* Wojciech Szmyd
* Paweł Szulc (pawel_szulc@onet.pl)
* Jan Twardowski
* Łukasz Wysogląd
* Kacper Zych (kacperzych@wp.pl)


## Disclaimer

This library is free software; you can redistribute it and/or
modify it under the terms of the GNU Lesser General Public
License as published by the Free Software Foundation; either
version 2.1 of the License, or (at your option) any later version.

This library is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public
License along with this library; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA


ChangeLog
=========================================
Software Defined Radio - DAB+ (libsdrdab)
=========================================

Version. 3.0, 2017-09-23
=============================
    - Remove custom RTL-SDR, now using system librtlsdr0
    - Refactoring fs and fc drift support (need more testing and tuning)
    - Support input RAW IQ files with formats: (u)int8, (u)int16, float, double, raw (uint8)
    - Custom resampler (RF IQ data) of type NN and Linear, code integrated as Resampler
    - Custom implementation of ReedSolomon decoder
    - Matlab model of various distortions (distort IQ samples for development purpose)
    - XPAD implementation, text only (need more development and integration)
    - Refactoring convolutional decoder subsystem, ready for different implementation nad real-time switching
    - Initial implementation of spiral convolutional decoder
    - Fix MPEG2 support (classic DAB)
    - Fix error in decoding EEP-B and EEP-A protection levels
    - Decode data from NULL
    - Acquire testing data
    - Generate mux testing data (DAB/DAB+, all protection levels, X-pad, etc...)
    - Improve synchronizer (cleaner, faster code)
    - Framework for extensive test of bitstream-level decoding for various mux configuration
    - Improve UnitTest for depuncturer (supports all combination)
    - Fix, improve and extend sdrdab-cli reference implementation
    - Fix synchronization audio player with data decoder (audio buffer is no longer overloaded)
    - Ubuntu package generator (need review and testing)

Version. 2.0, 2016-06-07
=============================
    - new Scheduler: faster, more stable, readable code
    - refactored DataDecoder
    - rewritten remodulation (fc): LUT-based
    - Viterbi optimization by means of SSE3 (up to 3x improvement)
    - refactored Synchronizer
    - support legacy DAB (MPEG) and DAB+ (AAC)
    - add two version of SNR calculation: from spectrum (RAW) and prefix
    - rewritten build system, now using cmake instead of Eclipse
    - MacOS X support
    - improved debug logging 
    - refactoring DataFeeder (need even more refactoring)
    - simple GUI implementation (need to be rewritten and improved)
    - acquired DAB+ signal for testing
    - upgraded sdrdab-cli (more parameters)
    - Matlab reference software for DAB(+) decoder
    - included unittest
    - MODE III still not supported    

Version. 1.0 BETA, 2015-07-07
=============================
    - x86, real-time implementation (i7-3520M: 20-100% CPU utilization depends on configuration)
    - decoding type: file2file, file2speaker, air2file, air2speaker
    - MODE I, MODE II, MODE IV (MODE III supported but buggy - do not works)
    - input devices: all tuners supported by sdrlib (tested on RLT2832U + R820T, R820T2)
    - input file: biarny, 8 bits per I/Q sample (unsigned int, 127 == 0), format: IQIQIQIQIQ...
    - change carrier frequency on air2* mode
    - fine tune carrier frequency (ppm)
    - 2.048 MHz base sampling frequency (resampling if necessary in file2* mode)
    - decode station list with names
    - switch between station in real-time
    - DAB+ (AAC) without XPAD, only audio (old DAB MPEG supported but buggy - do not works)


matlab/distort.m
% Help message
%
% DISTORT function created to set awgn, fs and fc drifts to clear recording
% usefull to distort recording in controlled way
% allows to check synchronization time in different scenarios
%
% distort(filename_r, filename_w, struct_table)
% Filename_w saved file uint8 0-255
% filename_w in .raw format
% filename_r in .dat format 'float' or .raw format 'uint8'
%
% You can create many different time intervals by adding structures to
% struct_table with variable settings
% struct_table = [data1 data2 ... dataN];
% dataN=struct('fs_drift',[],'fc_drift',[],'snr',[],'time_begin',[],'time_end',[]);
%
% fs_drift - set fs drift in Hz
% fc_drift - set fc drift in Hz
% snr - add awgn with snr level provided and measured
% time_begin - in which moment set distortion
% time_end - in which moment remove distortion
% fs is set to 2048e3 you can change it at the beggining of the function code
%
% Example:
% data1 = struct('fs_drift',[10],'fc_drift',[10],'snr',[10],'time_begin',[7],'time_end',[30]);
% data2 = struct('fs_drift',[10],'fc_drift',[10],'snr',[0],'time_begin',[10],'time_end',[25]);
% data3 = struct('fs_drift',[10],'fc_drift',[10],'snr',[0],'time_begin',[14],'time_end',[20]);
% enter_table=[data1 data2 data3];
% Filename_w=distort(example.dat, example.raw, enter_table);
%
% using fs_drift or fc_drift in the same interval more than once will cause
% summarize drifts
% in example above in interval 14-20 drift is set to 30, in intervals 10-14
% and 20-25 is set to 20 etc.
% in case of any questions reach me on email milosz.sliwinski96@gmail.com
%
% For more information, see <a href="matlab:
% web('http://www.mathworks.com')">the MathWorks Web site</a>.

if(0) %should be set to 0 unless you want to use test data
    %test data
    clear all
    close all
    time=10;
    
    filename_r = 'Record3_katowice_iq.dat'; %test value -- sdat=519045120 bytes
    filename_w = 'Record3_katowice_iq_test.raw'; %test value -- 129,761,280 bytes
    %data = struct('fs_drift',[],'fc_drift',[],'snr',[],'time_begin',[],'time_end',[]);
    %struct_table=[data data data];
    
    count=3;
    data1 = struct('fs_drift',[50],'fc_drift',[50],'snr',[10],'time_begin',[4],'time_end',[26]);
    data2 = struct('fs_drift',[50],'fc_drift',[50],'snr',[10],'time_begin',[8],'time_end',[22]);
    data3 = struct('fs_drift',[50],'fc_drift',[50],'snr',[10],'time_begin',[12],'time_end',[18]);
    enter_table=[data1 data2 data3];
    
    struct_table=[];
    for i=1:count
        if(floor(enter_table(i).time_end/time)-floor(enter_table(i).time_begin/time))
            k=floor(enter_table(i).time_end/time)-floor(enter_table(i).time_begin/time);
            %time_end=enter_table(i).time_begin+mod(-enter_table(i).time_begin,10);
            for j=0:k
                if(~j)
                    struct_element = struct('fs_drift',enter_table(i).fs_drift,'fc_drift',enter_table(i).fc_drift,'snr',enter_table(i).snr,'time_begin',enter_table(i).time_begin,'time_end',enter_table(i).time_begin+mod(-enter_table(i).time_begin,time));
                elseif(j && k-j)
                    struct_element = struct('fs_drift',enter_table(i).fs_drift,'fc_drift',enter_table(i).fc_drift,'snr',enter_table(i).snr,'time_begin',enter_table(i).time_begin+mod(-enter_table(i).time_begin,time)+(j-1)*time,'time_end',enter_table(i).time_begin+mod(-enter_table(i).time_begin,time)+j*time);
                elseif(~(k-j))
                    struct_element = struct('fs_drift',enter_table(i).fs_drift,'fc_drift',enter_table(i).fc_drift,'snr',enter_table(i).snr,'time_begin',enter_table(i).time_begin+mod(-enter_table(i).time_begin,time)+(j-1)*time,'time_end',enter_table(i).time_end);
                end %if
                struct_table = [struct_table struct_element];
            end %for
        end %if
    end %for
    
end %if() test data

%fs,fc drifts in Hz
function distort(filename_r, filename_w, enter_table)
good_data_flag = 0;
pattern1 = regexptranslate('wildcard','*.dat');
pattern2 = regexptranslate('wildcard','*.raw');
if(regexp(filename_r,pattern1))
    good_data_flag = 1;
    file_type='dat';
end %if
if(regexp(filename_r,pattern2))
    good_data_flag = 1;
    file_type='raw';
end %if
count=size(enter_table,2);
% table(1).fs_drift
%if value = 0 than there is no change
%max file lenght = 10s
time = 10; %in seconds
fs = 2048e3; %fs used
%each read sample is 'double' 64 (8)
len = fs*time; %~2e7

Filename = fopen(filename_r, 'rb');
Filename_w = fopen(filename_w, 'wb');
NSamples = [2, len]; %original NSamples=6e9
size_dat=dir(filename_r);
if file_type=='dat'
    file_size = size_dat.bytes/fs/time/8;
elseif file_type=='raw'
    file_size = size_dat.bytes/fs/time/8*4;
end %if
clear size_dat;
%check if time_end values meets file length
for i=1:count
    checking=floor(file_size*10);
    if(enter_table(i).time_end>checking)
        good_data_flag=0;
    end %if
end % for checking
struct_table=[];
%divide enter_table into smaller struct_tables
if(good_data_flag)
    for i=1:count
        if(floor(enter_table(i).time_end/time)-floor(enter_table(i).time_begin/time))
            k=floor(enter_table(i).time_end/time)-floor(enter_table(i).time_begin/time);
            %time_end=enter_table(i).time_begin+mod(-enter_table(i).time_begin,10);
            for j=0:k
                if(~j)
                    struct_element = struct('fs_drift',enter_table(i).fs_drift,'fc_drift',enter_table(i).fc_drift,'snr',enter_table(i).snr,'time_begin',enter_table(i).time_begin,'time_end',enter_table(i).time_begin+mod(-enter_table(i).time_begin,time));
                elseif(j && k-j)
                    struct_element = struct('fs_drift',enter_table(i).fs_drift,'fc_drift',enter_table(i).fc_drift,'snr',enter_table(i).snr,'time_begin',enter_table(i).time_begin+mod(-enter_table(i).time_begin,time)+(j-1)*time,'time_end',enter_table(i).time_begin+mod(-enter_table(i).time_begin,time)+j*time);
                elseif(~(k-j))
                    struct_element = struct('fs_drift',enter_table(i).fs_drift,'fc_drift',enter_table(i).fc_drift,'snr',enter_table(i).snr,'time_begin',enter_table(i).time_begin+mod(-enter_table(i).time_begin,time)+(j-1)*time,'time_end',enter_table(i).time_end);
                end %if
                struct_table = [struct_table struct_element];
            end %for
        end %if
    end %for
    
    
    counter=size(struct_table,2);
    file_size=floor(file_size);
    for i=0:1:file_size
        
        %20480000
        %16220160
        if file_type=='dat'
            %normalizing vector t and set mean value to 0
            t = fread(Filename, NSamples, 'float'); %327680000 - number of bytes read
            t = t-(mean(mean(t')));
            tmax = max(max(t'));
            tmin = min(min(t'));
            t = t./(max(tmax,-tmin));
            %one value, one sample
        elseif file_type=='raw'
            %normalizing vector t and set mean value to 0
            t = fread(Filename, NSamples, 'uint8'); %327680000/4 - number of bytes read
            t = t-(mean(mean(t')));
            tmax = max(max(t'));
            tmin = min(min(t'));
            t = t./(max(tmax,-tmin));
            %one value, one sample
        end
        %t(1,:) - real t(2,:) - imag
        %change fc
        for j=1:counter
            if(struct_table(j).fc_drift && ~floor(struct_table(j).time_begin/time-i))
                if(~mod(struct_table(j).time_end,time))
                    change_fragment=t(1,mod(struct_table(j).time_begin,time)*fs+1:10*fs)+1i*t(2,mod(struct_table(j).time_begin,time)*fs+1:10*fs);
                    %rotate now
                    rotated=[1:1:length(change_fragment)];
                    rotated=exp(1i*2*pi*struct_table(j).fc_drift/fs*rotated);
                    rotated=change_fragment.*rotated;
                    t(1,mod(struct_table(j).time_begin,time)*fs+1:10*fs)=real(rotated);
                    t(2,mod(struct_table(j).time_begin,time)*fs+1:10*fs)=imag(rotated);
                else
                    change_fragment=t(1,mod(struct_table(j).time_begin,time)*fs+1:mod(struct_table(j).time_end,time)*fs)+1i*t(2,mod(struct_table(j).time_begin,time)*fs+1:mod(struct_table(j).time_end,time)*fs);
                    %rotate now
                    rotated=[1:1:length(change_fragment)];
                    rotated=exp(1i*2*pi*struct_table(j).fc_drift/fs*rotated);
                    rotated=change_fragment.*rotated;
                    t(1,mod(struct_table(j).time_begin,time)*fs+1:mod(struct_table(j).time_end,time)*fs)=real(rotated);
                    t(2,mod(struct_table(j).time_begin,time)*fs+1:mod(struct_table(j).time_end,time)*fs)=imag(rotated);
                end %if
            end %if
        end %for
        %AWGN
        for j=1:counter
            if(struct_table(j).snr && ~floor(struct_table(j).time_begin/time-i))
                if(~mod(struct_table(j).time_end,time))
                    change_fragment=t(1,mod(struct_table(j).time_begin,time)*fs+1:10*fs)+1i*t(2,mod(struct_table(j).time_begin,time)*fs+1:10*fs);
                    change_fragment=awgn(change_fragment,struct_table(j).snr, 'measured');
                    t(1,mod(struct_table(j).time_begin,time)*fs+1:10*fs)=real(change_fragment);
                    t(2,mod(struct_table(j).time_begin,time)*fs+1:10*fs)=imag(change_fragment);
                else
                    change_fragment=t(1,mod(struct_table(j).time_begin,time)*fs+1:mod(struct_table(j).time_end,time)*fs)+1i*t(2,mod(struct_table(j).time_begin,time)*fs+1:mod(struct_table(j).time_end,time)*fs);
                    change_fragment=awgn(change_fragment,struct_table(j).snr, 'measured');
                    t(1,mod(struct_table(j).time_begin,time)*fs+1:mod(struct_table(j).time_end,time)*fs)=real(change_fragment);
                    t(2,mod(struct_table(j).time_begin,time)*fs+1:mod(struct_table(j).time_end,time)*fs)=imag(change_fragment);
                end %if
            end %if
        end %for
        %resample
        %interpol here
        for j=1:counter
            if(struct_table(j).fs_drift && ~floor(struct_table(j).time_begin/time-i))
                if(~mod(struct_table(j).time_end,time))
                    change_fragment=t(:,mod(struct_table(j).time_begin,time)*fs+1:10*fs);
                    %x=[1:1:length(change_fragment)];
                    %xq=[1:fs/(fs+fs_drift):length(change_fragment)];
                    %interp1(x,v,xq,method),
                    change_fragment = interp1([1:1:length(change_fragment)],change_fragment',[1:fs/(fs+struct_table(j).fs_drift):length(change_fragment)],'spline');
                    t=[t(:,1:mod(struct_table(j).time_begin,time)*fs+1) change_fragment' t(:,struct_table(j).time_end*fs+1:end)];
                else
                    change_fragment=t(:,mod(struct_table(j).time_begin,time)*fs+1:mod(struct_table(j).time_end,time)*fs);
                    %x=[1:1:length(change_fragment)];
                    %xq=[1:fs/(fs+fs_drift):length(change_fragment)];
                    %interp1(x,v,xq,method),
                    change_fragment = interp1([1:1:length(change_fragment)],change_fragment',[1:fs/(fs+struct_table(j).fs_drift):length(change_fragment)],'spline');
                    t=[t(:,1:mod(struct_table(j).time_begin,time)*fs+1) change_fragment' t(:,mod(struct_table(j).time_end,time)*fs+1:end)];
                end %if
            end %if
        end %for
        %define one column vector b
        b = zeros(1, 2*length(t));
        %odd values are real
        b(1:2:2*length(t))=t(1,:);
        %even values are imag
        b(2:2:2*length(t))=t(2,:);
        
        %convert normalized float values to uint in range 0-255
        b = 128 * b + 127;
        b = uint8(b);
        
        fwrite( Filename_w, b, 'uint8' );
    end %for
    fclose(Filename);
    fclose(Filename_w);
end %if(good_data_flag)
end %function